File: .skills/openclaw-skills/skills/0xnyk/xint/lib/media.ts
/**
 * media.ts — Download media from a tweet using official X API endpoints.
 */

import { mkdirSync, writeFileSync } from "fs";
import { join, resolve } from "path";
import { BASE, bearerGet } from "./api";
import { trackCost } from "./costs";

interface RawMediaVariant {
  bit_rate?: number;
  content_type?: string;
  url?: string;
}

interface RawMedia {
  media_key: string;
  type: "photo" | "video" | "animated_gif" | string;
  url?: string;
  preview_image_url?: string;
  variants?: RawMediaVariant[];
}

interface DownloadRecord {
  mediaKey: string;
  type: string;
  sourceUrl?: string;
  savedPath?: string;
  bytes?: number;
  error?: string;
}

type MediaFilter = "all" | "photos" | "video";
interface FileNameContext {
  tweetId: string;
  username?: string;
  index: number; // 1-based
  mediaType: string;
  mediaKey: string;
  createdAt?: string;
  ext: string;
}

const DOWNLOAD_ATTEMPTS = 3;
const BACKOFF_BASE_MS = 500;
const DEFAULT_NAME_TEMPLATE = "{tweet_id}-{index}-{type}";

export function extractTweetId(input: string): string | null {
  const trimmed = input.trim();
  if (/^\d+$/.test(trimmed)) return trimmed;

  const match = trimmed.match(
    /(?:x|twitter)\.com\/(?:(?:[^/\s]+\/status)|(?:i\/web\/status)|(?:i\/status))\/(\d+)/i,
  );
  if (match) return match[1];

  const fallback = trimmed.match(/status\/(\d+)/i);
  return fallback ? fallback[1] : null;
}

export function inferExtension(mediaUrl: string, mediaType: string): string {
  try {
    const parsed = new URL(mediaUrl);
    const pathname = parsed.pathname.toLowerCase();
    const dot = pathname.lastIndexOf(".");
    if (dot > 0) {
      const ext = pathname.slice(dot + 1);
      if (/^[a-z0-9]{2,5}$/.test(ext)) return ext;
    }

    const format = parsed.searchParams.get("format");
    if (format && /^[a-z0-9]{2,5}$/i.test(format)) return format.toLowerCase();
  } catch {
    // Fall through to type-based defaults.
  }

  if (mediaType === "video" || mediaType === "animated_gif") return "mp4";
  return "jpg";
}

export function parseMaxItems(value: string | undefined): number | null {
  if (!value) return null;
  const n = Number.parseInt(value, 10);
  if (!Number.isFinite(n) || n <= 0) return null;
  return n;
}

function isVideoLike(type: string): boolean {
  return type === "video" || type === "animated_gif";
}

export function matchesMediaFilter(media: RawMedia, filter: MediaFilter): boolean {
  if (filter === "all") return true;
  if (filter === "photos") return media.type === "photo";
  return isVideoLike(media.type);
}

export function selectDownloadUrl(media: RawMedia): string | null {
  if (media.type === "photo") {
    return media.url || media.preview_image_url || null;
  }

  const variants = (media.variants || [])
    .filter((v) => v.url && (v.content_type || "").toLowerCase().includes("mp4"))
    .sort((a, b) => (b.bit_rate || 0) - (a.bit_rate || 0));

  if (variants.length > 0) return variants[0].url || null;
  return media.preview_image_url || media.url || null;
}

function sanitizeNamePart(value: string): string {
  return value.replace(/[^a-zA-Z0-9_-]+/g, "_").replace(/_+/g, "_").replace(/^_+|_+$/g, "");
}

function sanitizeFileName(value: string): string {
  return value
    .replace(/[^a-zA-Z0-9._-]+/g, "_")
    .replace(/_+/g, "_")
    .replace(/^_+|_+$/g, "");
}

export function renderFileNameTemplate(
  template: string | undefined,
  ctx: FileNameContext,
): string {
  const rawTemplate = template?.trim() || DEFAULT_NAME_TEMPLATE;
  const replacements: Record<string, string> = {
    "{tweet_id}": sanitizeNamePart(ctx.tweetId),
    "{username}": sanitizeNamePart(ctx.username || "unknown"),
    "{index}": String(ctx.index),
    "{type}": sanitizeNamePart(ctx.mediaType || "media"),
    "{media_key}": sanitizeNamePart(ctx.mediaKey || "unknown"),
    "{created_at}": sanitizeNamePart(ctx.createdAt || "unknown"),
    "{ext}": sanitizeNamePart(ctx.ext || "bin"),
  };

  let rendered = rawTemplate;
  for (const [token, value] of Object.entries(replacements)) {
    rendered = rendered.split(token).join(value);
  }

  let base = sanitizeFileName(rendered);
  if (!base) {
    base = `${sanitizeNamePart(ctx.tweetId)}-${ctx.index}-${sanitizeNamePart(ctx.mediaType || "media")}`;
  }

  const hasExt = /\.[a-zA-Z0-9]{2,5}$/.test(base);
  return hasExt ? base : `${base}.${ctx.ext}`;
}

function defaultMediaDir(): string {
  return join(import.meta.dir, "..", "data", "media");
}

function parseFlags(args: string[]): {
  target?: string;
  outDir?: string;
  maxItems?: number;
  nameTemplate?: string;
  mediaFilter: MediaFilter;
  asJson: boolean;
  help: boolean;
} {
  const positional: string[] = [];
  let outDir: string | undefined;
  let maxItems: number | undefined;
  let nameTemplate: string | undefined;
  let mediaFilter: MediaFilter = "all";
  let asJson = false;
  let help = false;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    switch (arg) {
      case "--dir":
        outDir = args[++i];
        break;
      case "--max-items": {
        const parsed = parseMaxItems(args[++i]);
        if (parsed === null) {
          throw new Error("Usage: --max-items requires a positive integer.");
        }
        maxItems = parsed;
        break;
      }
      case "--name-template":
        nameTemplate = args[++i];
        if (!nameTemplate || !nameTemplate.trim()) {
          throw new Error("Usage: --name-template requires a non-empty value.");
        }
        break;
      case "--photos-only":
        if (mediaFilter === "video") {
          throw new Error("Use only one of --photos-only or --video-only.");
        }
        mediaFilter = "photos";
        break;
      case "--video-only":
        if (mediaFilter === "photos") {
          throw new Error("Use only one of --photos-only or --video-only.");
        }
        mediaFilter = "video";
        break;
      case "--json":
        asJson = true;
        break;
      case "--help":
      case "-h":
        help = true;
        break;
      default:
        positional.push(arg);
        break;
    }
  }

  return { target: positional[0], outDir, maxItems, nameTemplate, mediaFilter, asJson, help };
}

async function sleepMs(ms: number): Promise<void> {
  await new Promise((resolveFn) => setTimeout(resolveFn, ms));
}

function shouldRetryStatus(status: number): boolean {
  return status === 408 || status === 425 || status === 429 || status >= 500;
}

async function downloadBytesWithRetry(url: string): Promise<Buffer> {
  for (let attempt = 1; attempt <= DOWNLOAD_ATTEMPTS; attempt++) {
    try {
      const res = await fetch(url);
      if (res.ok) {
        return Buffer.from(await res.arrayBuffer());
      }

      const canRetry = shouldRetryStatus(res.status) && attempt < DOWNLOAD_ATTEMPTS;
      if (!canRetry) {
        throw new Error(`HTTP ${res.status}`);
      }
    } catch (err: any) {
      if (attempt >= DOWNLOAD_ATTEMPTS) {
        throw err;
      }
    }

    const delayMs = BACKOFF_BASE_MS * 2 ** (attempt - 1);
    await sleepMs(delayMs);
  }

  throw new Error("Download failed after retries.");
}

function printHelp(): void {
  console.log(`
Usage: xint media <tweet_id|tweet_url> [options]

Download media attachments from a tweet using official X API metadata.

Options:
  --dir <path>       Output directory (default: data/media)
  --max-items <N>    Download up to N media items
  --name-template    Filename template (tokens: {tweet_id} {username} {index}
                     {type} {media_key} {created_at} {ext})
  --photos-only      Download photos only
  --video-only       Download videos/GIFs only
  --json             Output JSON summary
  --help, -h         Show this help

Examples:
  xint media 1900100012345678901
  xint media https://x.com/user/status/1900100012345678901
  xint media 1900100012345678901 --name-template "{username}-{created_at}-{index}"
  xint media 1900100012345678901 --video-only --max-items 1
  xint media 1900100012345678901 --dir ./downloads --json
`);
}

export async function cmdMedia(args: string[]): Promise<void> {
  const { target, outDir, maxItems, nameTemplate, mediaFilter, asJson, help } = parseFlags(args);
  if (help || !target) {
    printHelp();
    return;
  }

  const tweetId = extractTweetId(target);
  if (!tweetId) {
    throw new Error("Invalid tweet ID or tweet URL.");
  }

  const outputDir = resolve(outDir || defaultMediaDir());
  mkdirSync(outputDir, { recursive: true });

  const query =
    "expansions=attachments.media_keys,author_id&tweet.fields=attachments,author_id,created_at&user.fields=username&media.fields=type,url,preview_image_url,variants,width,height,duration_ms,alt_text";
  const url = `${BASE}/tweets/${tweetId}?${query}`;
  const raw = await bearerGet(url);
  trackCost("media_metadata", `/2/tweets/${tweetId}`, 1);

  const mediaList: RawMedia[] = Array.isArray((raw as any)?.includes?.media)
    ? (raw as any).includes.media
    : [];
  const users = Array.isArray((raw as any)?.includes?.users) ? (raw as any).includes.users : [];
  const authorId = (raw as any)?.data?.author_id;
  const createdAt = typeof (raw as any)?.data?.created_at === "string" ? (raw as any).data.created_at : undefined;
  const username = users.find((u: any) => u.id === authorId)?.username || null;

  if (mediaList.length === 0) {
    console.log("No media attachments found on this tweet.");
    return;
  }

  const selectedMedia = mediaList.filter((m) => matchesMediaFilter(m, mediaFilter));
  const limitedMedia = maxItems ? selectedMedia.slice(0, maxItems) : selectedMedia;
  if (limitedMedia.length === 0) {
    console.log("No media attachments matched the selected filters.");
    return;
  }

  const records: DownloadRecord[] = [];

  for (let i = 0; i < limitedMedia.length; i++) {
    const media = limitedMedia[i];
    const sourceUrl = selectDownloadUrl(media);
    const baseRecord: DownloadRecord = {
      mediaKey: media.media_key,
      type: media.type,
      sourceUrl: sourceUrl || undefined,
    };

    if (!sourceUrl) {
      records.push({ ...baseRecord, error: "No downloadable URL found" });
      continue;
    }

    try {
      const bytes = await downloadBytesWithRetry(sourceUrl);
      const ext = inferExtension(sourceUrl, media.type);
      const fileName = renderFileNameTemplate(nameTemplate, {
        tweetId,
        username: username || undefined,
        index: i + 1,
        mediaType: media.type || "media",
        mediaKey: media.media_key || "unknown",
        createdAt,
        ext,
      });
      const path = join(outputDir, fileName);
      writeFileSync(path, bytes);

      records.push({
        ...baseRecord,
        savedPath: path,
        bytes: bytes.length,
      });
    } catch (err: any) {
      records.push({
        ...baseRecord,
        error: err?.message || String(err),
      });
    }
  }

  if (asJson) {
    console.log(
      JSON.stringify(
        {
          tweetId,
          username,
          outputDir,
          totalMedia: mediaList.length,
          selectedMedia: limitedMedia.length,
          nameTemplate: nameTemplate || DEFAULT_NAME_TEMPLATE,
          downloaded: records.filter((r) => !r.error).length,
          records,
        },
        null,
        2,
      ),
    );
    return;
  }

  console.log(`\n🎞️  Media download for tweet ${tweetId}${username ? ` (@${username})` : ""}`);
  console.log(`Output directory: ${outputDir}`);
  console.log(`Attachments found: ${mediaList.length} (selected: ${limitedMedia.length})\n`);

  for (let i = 0; i < records.length; i++) {
    const rec = records[i];
    if (rec.error) {
      console.log(`${i + 1}. ❌ ${rec.type} (${rec.mediaKey}) — ${rec.error}`);
      continue;
    }
    console.log(`${i + 1}. ✅ ${rec.type} (${rec.mediaKey})`);
    console.log(`   ${rec.savedPath} (${rec.bytes} bytes)`);
  }

  const success = records.filter((r) => !r.error).length;
  const failed = records.length - success;
  console.log(`\nDone: ${success} downloaded, ${failed} failed.`);
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/media.rs
use anyhow::Result;
use serde::Serialize;
use std::fs;
use std::path::PathBuf;
use std::time::Duration;

use crate::cli::MediaArgs;
use crate::client::XClient;
use crate::config::Config;
use crate::costs;
use crate::format;

#[derive(Debug, Serialize)]
struct DownloadRecord {
    media_key: String,
    media_type: String,
    source_url: Option<String>,
    saved_path: Option<String>,
    bytes: Option<usize>,
    error: Option<String>,
}

#[derive(Clone, Copy)]
enum MediaFilter {
    All,
    Photos,
    Video,
}

const DOWNLOAD_ATTEMPTS: u8 = 3;
const BACKOFF_BASE_MS: u64 = 500;
const DEFAULT_NAME_TEMPLATE: &str = "{tweet_id}-{index}-{type}";

struct FileNameContext<'a> {
    tweet_id: &'a str,
    username: Option<&'a str>,
    index: usize,
    media_type: &'a str,
    media_key: &'a str,
    created_at: Option<&'a str>,
    ext: &'a str,
}

pub async fn run(args: &MediaArgs, config: &Config, client: &XClient) -> Result<()> {
    if args.photos_only && args.video_only {
        anyhow::bail!("Use only one of --photos-only or --video-only.");
    }
    if let Some(max_items) = args.max_items {
        if max_items == 0 {
            anyhow::bail!("--max-items must be a positive integer.");
        }
    }
    if let Some(tpl) = &args.name_template {
        if tpl.trim().is_empty() {
            anyhow::bail!("--name-template must be a non-empty value.");
        }
    }

    let token = config.require_bearer_token()?;
    let tweet_id = extract_tweet_id(&args.target).ok_or_else(|| {
        anyhow::anyhow!("Invalid tweet ID or tweet URL. Expected numeric ID or status URL.")
    })?;
    let out_dir = resolve_output_dir(args.dir.as_ref(), config)?;

    let path = format!(
        "tweets/{tweet_id}?expansions=attachments.media_keys,author_id&tweet.fields=attachments,author_id,created_at&user.fields=username&media.fields=type,url,preview_image_url,variants,width,height,duration_ms,alt_text"
    );
    let raw = client.bearer_get(&path, token).await?;
    costs::track_cost(
        &config.costs_path(),
        "media_metadata",
        &format!("/2/tweets/{tweet_id}"),
        1,
    );

    let media_items = raw
        .includes
        .as_ref()
        .and_then(|i| i.media.as_ref())
        .cloned()
        .unwrap_or_default();
    let total_media = media_items.len();

    if total_media == 0 {
        println!("No media attachments found on this tweet.");
        return Ok(());
    }

    let media_filter = if args.photos_only {
        MediaFilter::Photos
    } else if args.video_only {
        MediaFilter::Video
    } else {
        MediaFilter::All
    };
    let mut selected_media: Vec<serde_json::Value> = media_items
        .into_iter()
        .filter(|m| media_matches_filter(m, media_filter))
        .collect();
    if let Some(max_items) = args.max_items {
        selected_media.truncate(max_items);
    }
    if selected_media.is_empty() {
        println!("No media attachments matched the selected filters.");
        return Ok(());
    }

    let author_id = raw
        .data
        .as_ref()
        .and_then(|d| d.get("author_id"))
        .and_then(|v| v.as_str())
        .unwrap_or_default();
    let created_at = raw
        .data
        .as_ref()
        .and_then(|d| d.get("created_at"))
        .and_then(|v| v.as_str())
        .map(ToString::to_string);
    let username = raw
        .includes
        .as_ref()
        .and_then(|i| i.users.as_ref())
        .and_then(|users| {
            users
                .iter()
                .find(|u| u.id == author_id)
                .and_then(|u| u.username.clone())
        });

    let http = reqwest::Client::new();
    let mut records: Vec<DownloadRecord> = Vec::new();

    for (index, media) in selected_media.iter().enumerate() {
        let media_key = media
            .get("media_key")
            .and_then(|v| v.as_str())
            .unwrap_or("unknown")
            .to_string();
        let media_type = media
            .get("type")
            .and_then(|v| v.as_str())
            .unwrap_or("media")
            .to_string();
        let source_url = select_download_url(media);

        let base = DownloadRecord {
            media_key: media_key.clone(),
            media_type: media_type.clone(),
            source_url: source_url.clone(),
            saved_path: None,
            bytes: None,
            error: None,
        };

        let Some(url) = source_url else {
            records.push(DownloadRecord {
                error: Some("No downloadable URL found".to_string()),
                ..base
            });
            continue;
        };

        let bytes = match download_bytes_with_retry(&http, &url).await {
            Ok(b) => b,
            Err(err) => {
                records.push(DownloadRecord {
                    error: Some(err.to_string()),
                    ..base
                });
                continue;
            }
        };

        let ext = infer_extension(&url, &media_type);
        let file_name = build_file_name(
            args.name_template.as_deref(),
            &FileNameContext {
                tweet_id: &tweet_id,
                username: username.as_deref(),
                index: index + 1,
                media_type: &media_type,
                media_key: &media_key,
                created_at: created_at.as_deref(),
                ext: &ext,
            },
        );
        let file_path = out_dir.join(file_name);
        if let Err(err) = fs::write(&file_path, &bytes) {
            records.push(DownloadRecord {
                error: Some(err.to_string()),
                ..base
            });
            continue;
        }

        records.push(DownloadRecord {
            saved_path: Some(file_path.to_string_lossy().to_string()),
            bytes: Some(bytes.len()),
            ..base
        });
    }

    if args.json {
        let payload = serde_json::json!({
            "tweet_id": tweet_id,
            "username": username,
            "output_dir": out_dir.to_string_lossy(),
            "total_media": total_media,
            "selected_media": selected_media.len(),
            "name_template": args.name_template.as_deref().unwrap_or(DEFAULT_NAME_TEMPLATE),
            "downloaded": records.iter().filter(|r| r.error.is_none()).count(),
            "records": records,
        });
        format::print_json_pretty_filtered(&payload)?;
        return Ok(());
    }

    println!(
        "\n🎞️  Media download for tweet {tweet_id}{}",
        username
            .as_ref()
            .map(|u| format!(" (@{u})"))
            .unwrap_or_default()
    );
    println!("Output directory: {}", out_dir.to_string_lossy());
    println!(
        "Attachments found: {} (selected: {})\n",
        total_media,
        selected_media.len()
    );

    for (idx, rec) in records.iter().enumerate() {
        if let Some(err) = &rec.error {
            println!(
                "{}. ❌ {} ({}) — {}",
                idx + 1,
                rec.media_type,
                rec.media_key,
                err
            );
        } else {
            println!("{}. ✅ {} ({})", idx + 1, rec.media_type, rec.media_key);
            if let (Some(path), Some(size)) = (&rec.saved_path, rec.bytes) {
                println!("   {path} ({size} bytes)");
            }
        }
    }

    let success = records.iter().filter(|r| r.error.is_none()).count();
    let failed = records.len().saturating_sub(success);
    println!("\nDone: {success} downloaded, {failed} failed.");

    Ok(())
}

fn is_video_like(media_type: &str) -> bool {
    media_type == "video" || media_type == "animated_gif"
}

fn sanitize_file_name(value: &str) -> String {
    let mut out = String::new();
    let mut last_was_underscore = false;
    for ch in value.chars() {
        let valid = ch.is_ascii_alphanumeric() || ch == '-' || ch == '_' || ch == '.';
        if valid {
            out.push(ch);
            last_was_underscore = false;
        } else if !last_was_underscore {
            out.push('_');
            last_was_underscore = true;
        }
    }
    out.trim_matches('_').to_string()
}

fn has_extension(value: &str) -> bool {
    if let Some(dot) = value.rfind('.') {
        let ext = &value[dot + 1..];
        return (2..=5).contains(&ext.len()) && ext.chars().all(|c| c.is_ascii_alphanumeric());
    }
    false
}

fn build_file_name(template: Option<&str>, ctx: &FileNameContext<'_>) -> String {
    let raw_template = template
        .map(str::trim)
        .filter(|tpl| !tpl.is_empty())
        .unwrap_or(DEFAULT_NAME_TEMPLATE);

    let mut rendered = raw_template.to_string();
    let replacements = [
        ("{tweet_id}", sanitize_name_part(ctx.tweet_id)),
        (
            "{username}",
            sanitize_name_part(ctx.username.unwrap_or("unknown")),
        ),
        ("{index}", ctx.index.to_string()),
        ("{type}", sanitize_name_part(ctx.media_type)),
        ("{media_key}", sanitize_name_part(ctx.media_key)),
        (
            "{created_at}",
            sanitize_name_part(ctx.created_at.unwrap_or("unknown")),
        ),
        ("{ext}", sanitize_name_part(ctx.ext)),
    ];
    for (token, value) in replacements {
        rendered = rendered.replace(token, &value);
    }

    let mut base = sanitize_file_name(&rendered);
    if base.is_empty() {
        base = format!(
            "{}-{}-{}",
            sanitize_name_part(ctx.tweet_id),
            ctx.index,
            sanitize_name_part(ctx.media_type)
        );
    }

    if has_extension(&base) {
        base
    } else {
        format!("{}.{}", base, ctx.ext)
    }
}

fn media_matches_filter(media: &serde_json::Value, filter: MediaFilter) -> bool {
    let media_type = media
        .get("type")
        .and_then(|v| v.as_str())
        .unwrap_or_default();
    match filter {
        MediaFilter::All => true,
        MediaFilter::Photos => media_type == "photo",
        MediaFilter::Video => is_video_like(media_type),
    }
}

fn should_retry_status(status: u16) -> bool {
    status == 408 || status == 425 || status == 429 || status >= 500
}

async fn download_bytes_with_retry(http: &reqwest::Client, url: &str) -> Result<Vec<u8>> {
    let mut last_error = String::new();

    for attempt in 1..=DOWNLOAD_ATTEMPTS {
        match http.get(url).send().await {
            Ok(response) => {
                let status = response.status().as_u16();
                if response.status().is_success() {
                    let bytes = response.bytes().await?;
                    return Ok(bytes.to_vec());
                }

                let retryable = should_retry_status(status) && attempt < DOWNLOAD_ATTEMPTS;
                if !retryable {
                    anyhow::bail!("HTTP {status}");
                }
                last_error = format!("HTTP {status}");
            }
            Err(err) => {
                if attempt >= DOWNLOAD_ATTEMPTS {
                    return Err(err.into());
                }
                last_error = err.to_string();
            }
        }

        let delay_ms = BACKOFF_BASE_MS * (1_u64 << (attempt - 1));
        tokio::time::sleep(Duration::from_millis(delay_ms)).await;
    }

    anyhow::bail!("Download failed after retries: {last_error}")
}

fn resolve_output_dir(dir: Option<&String>, config: &Config) -> Result<PathBuf> {
    let base = match dir {
        Some(value) => {
            let p = PathBuf::from(value);
            if p.is_absolute() {
                p
            } else {
                std::env::current_dir()?.join(p)
            }
        }
        None => config.data_dir.join("media"),
    };

    fs::create_dir_all(&base)?;
    Ok(match base.canonicalize() {
        Ok(path) => path,
        Err(_) => base,
    })
}

fn extract_tweet_id(input: &str) -> Option<String> {
    let trimmed = input.trim();
    if trimmed.chars().all(|c| c.is_ascii_digit()) {
        return Some(trimmed.to_string());
    }

    let status_idx = trimmed.find("/status/")?;
    let id_part = &trimmed[status_idx + "/status/".len()..];
    let digits: String = id_part.chars().take_while(|c| c.is_ascii_digit()).collect();
    if digits.is_empty() {
        None
    } else {
        Some(digits)
    }
}

fn sanitize_name_part(value: &str) -> String {
    let mut out = String::new();
    let mut last_was_underscore = false;
    for ch in value.chars() {
        let valid = ch.is_ascii_alphanumeric() || ch == '-' || ch == '_';
        if valid {
            out.push(ch);
            last_was_underscore = false;
        } else if !last_was_underscore {
            out.push('_');
            last_was_underscore = true;
        }
    }
    let trimmed = out.trim_matches('_').to_string();
    if trimmed.is_empty() {
        "media".to_string()
    } else {
        trimmed
    }
}

fn infer_extension(source_url: &str, media_type: &str) -> String {
    if let Ok(url) = url::Url::parse(source_url) {
        let path = url.path().to_lowercase();
        if let Some(dot) = path.rfind('.') {
            let ext = &path[dot + 1..];
            if (2..=5).contains(&ext.len()) && ext.chars().all(|c| c.is_ascii_alphanumeric()) {
                return ext.to_string();
            }
        }

        if let Some(format) = url.query_pairs().find_map(|(k, v)| {
            if k == "format" {
                Some(v.to_string())
            } else {
                None
            }
        }) {
            let lower = format.to_lowercase();
            if (2..=5).contains(&lower.len()) && lower.chars().all(|c| c.is_ascii_alphanumeric()) {
                return lower;
            }
        }
    }

    if media_type == "video" || media_type == "animated_gif" {
        "mp4".to_string()
    } else {
        "jpg".to_string()
    }
}

fn select_download_url(media: &serde_json::Value) -> Option<String> {
    let media_type = media
        .get("type")
        .and_then(|v| v.as_str())
        .unwrap_or_default();
    if media_type == "photo" {
        return media
            .get("url")
            .and_then(|v| v.as_str())
            .map(ToString::to_string)
            .or_else(|| {
                media
                    .get("preview_image_url")
                    .and_then(|v| v.as_str())
                    .map(ToString::to_string)
            });
    }

    if let Some(variants) = media.get("variants").and_then(|v| v.as_array()) {
        let mut best_url: Option<String> = None;
        let mut best_rate: i64 = -1;
        for variant in variants {
            let content_type = variant
                .get("content_type")
                .and_then(|v| v.as_str())
                .unwrap_or_default()
                .to_lowercase();
            if !content_type.contains("mp4") {
                continue;
            }
            let Some(url) = variant.get("url").and_then(|v| v.as_str()) else {
                continue;
            };
            let bit_rate = variant
                .get("bit_rate")
                .and_then(|v| v.as_i64())
                .unwrap_or(0);
            if bit_rate > best_rate {
                best_rate = bit_rate;
                best_url = Some(url.to_string());
            }
        }
        if best_url.is_some() {
            return best_url;
        }
    }

    media
        .get("preview_image_url")
        .and_then(|v| v.as_str())
        .map(ToString::to_string)
        .or_else(|| {
            media
                .get("url")
                .and_then(|v| v.as_str())
                .map(ToString::to_string)
        })
}

#[cfg(test)]
mod tests {
    use super::{
        build_file_name, extract_tweet_id, infer_extension, media_matches_filter,
        select_download_url, should_retry_status, FileNameContext, MediaFilter,
    };

    #[test]
    fn extract_tweet_id_supports_id_and_url() {
        assert_eq!(
            extract_tweet_id("1900100012345678901"),
            Some("1900100012345678901".to_string())
        );
        assert_eq!(
            extract_tweet_id("https://x.com/user/status/1900100012345678901"),
            Some("1900100012345678901".to_string())
        );
        assert_eq!(extract_tweet_id("https://example.com"), None);
    }

    #[test]
    fn infer_extension_falls_back_by_media_type() {
        assert_eq!(
            infer_extension("https://pbs.twimg.com/media/abc", "photo"),
            "jpg"
        );
        assert_eq!(
            infer_extension("https://video.twimg.com/ext_tw_video/123", "video"),
            "mp4"
        );
    }

    #[test]
    fn select_download_url_prefers_highest_mp4_bitrate() {
        let media = serde_json::json!({
            "type": "video",
            "variants": [
                { "content_type": "video/mp4", "bit_rate": 256000, "url": "https://video/low.mp4" },
                { "content_type": "video/mp4", "bit_rate": 832000, "url": "https://video/high.mp4" }
            ]
        });
        assert_eq!(
            select_download_url(&media),
            Some("https://video/high.mp4".to_string())
        );
    }

    #[test]
    fn media_filter_matches_expected_types() {
        let photo = serde_json::json!({ "type": "photo" });
        let video = serde_json::json!({ "type": "video" });
        let gif = serde_json::json!({ "type": "animated_gif" });

        assert!(media_matches_filter(&photo, MediaFilter::Photos));
        assert!(!media_matches_filter(&video, MediaFilter::Photos));
        assert!(media_matches_filter(&video, MediaFilter::Video));
        assert!(media_matches_filter(&gif, MediaFilter::Video));
        assert!(!media_matches_filter(&photo, MediaFilter::Video));
    }

    #[test]
    fn retryable_statuses_are_identified() {
        assert!(should_retry_status(429));
        assert!(should_retry_status(503));
        assert!(!should_retry_status(404));
    }

    #[test]
    fn build_file_name_uses_placeholders_and_appends_ext() {
        let name = build_file_name(
            Some("{username}-{created_at}-{index}"),
            &FileNameContext {
                tweet_id: "1900100012345678901",
                username: Some("alice"),
                index: 2,
                media_type: "photo",
                media_key: "3_1",
                created_at: Some("2026-02-17T10:22:30.000Z"),
                ext: "jpg",
            },
        );
        assert_eq!(name, "alice-2026-02-17T10_22_30_000Z-2.jpg");
    }

    #[test]
    fn build_file_name_keeps_explicit_extension_placeholder() {
        let name = build_file_name(
            Some("{tweet_id}-{index}.{ext}"),
            &FileNameContext {
                tweet_id: "1900100012345678901",
                username: Some("alice"),
                index: 1,
                media_type: "video",
                media_key: "3_2",
                created_at: None,
                ext: "mp4",
            },
        );
        assert_eq!(name, "1900100012345678901-1.mp4");
    }
}
